lib/authn.js

```diff
@@ -67,86 +37,63 @@ export async function openDatabase(options) {
       `MongoDB server version "${stringVersion}" is unsupported.`,
       'NotSupportedError');
   }
+  // makes an unauthenticated call on the server
+  // to see if auth is required
+  const authRequired = await _authRequired({config, admin});
 
-  // try to authenticate...
-
-  // if there is a username & password set, create an auth object for Mongo;
+  // if authRequired create an auth object for Mongo;
   // otherwise `auth` will be passed as `null` and success will rely on other
-  // config options
-  let auth = null;
-  if(config.username && config.password) {
-    auth = {
+  // config options such as the url for the server.
+  if(authRequired) {
+    opts.connectOptions.auth = {
       user: config.username,
       password: config.password
     };
+    // authSource is the database to authenticate against
+    // this is usually `admin` in dev and a specific db in production
+    opts.connectOptions.authSource =
+      config.connectOptions.authSource || options.name;
   }
 
-  // authSource should be set in connectOptions
-  const opts = {
-    ...config.authentication,
-    ...config.connectOptions,
-    ...config.writeOptions,
-  };
-
   // if the user specified a connection URL use it
-  let url = config.url;
-  if(!url) {
-    url = urls.create(config);
+  if(!opts.url) {
```

Comment:
This doesn't seem right ... should it be?:

```suggestion
  if(config.url) {
```

Or maybe:

```suggestion
  if(!opts.url && config.url) {
```

Otherwise it seems like it deviates from previous behavior and / or may cause breakage if `config.url` is not present? Or is there some other kind of correction happening?

Reply:
ok so `openDatabase` copies the connection config stuff from config and passes those options to `_connect` now. There's a lot of clean up here.